lib/db/schema/patch-045-046.sql

```diff
@@ -0,0 +1,224 @@
+  isPrimary BOOLEAN NOT NULL DEFAULT FALSE,
+  verifiedAt BIGINT UNSIGNED,
+  createdAt BIGINT UNSIGNED NOT NULL,
+  INDEX `uid` (`uid`)
+) ENGINE=InnoDB;
+
+DROP procedure IF EXISTS `createEmail_1`;
+
+CREATE PROCEDURE `createEmail_1` (
+    IN `normalizedEmail` VARCHAR(255),
+    IN `email` VARCHAR(255),
+    IN `uid` BINARY(16) ,
+    IN `emailCode` BINARY(16),
+    IN `isVerified` TINYINT(1),
+    IN `isPrimary` TINYINT(1),
+    IN `verifiedAt` BIGINT UNSIGNED,
+    IN `createdAt` BIGINT UNSIGNED
+)
+BEGIN
+    -- Currently, can not add an email that is specified in the
+    -- accounts table, regardless of verification state.
+    SET @emailExists = 0;
+    SELECT COUNT(*) INTO @emailExists FROM accounts a WHERE a.normalizedEmail = normalizedEmail;
+    IF @emailExists > 0 THEN
+        SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO = 1062, MESSAGE_TEXT = 'Email already exists in accounts table.';
+    END IF;
+
+    INSERT INTO emails(
+        normalizedEmail,
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        verifiedAt,
+        createdAt
+    )
+    VALUES(
+        LOWER(normalizedEmail),
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        verifiedAt,
+        createdAt
+    );
+
+    COMMIT;
+END;
+
+CREATE PROCEDURE `accountEmails_1` (
+    IN `inUid` BINARY(16)
+)
+BEGIN
+    -- Return all emails that belong to the uid.
+    -- Email from the accounts table is considered the primary email.
+	(SELECT
+        a.normalizedEmail,
+        a.email,
+        a.uid,
+        a.emailCode,
+        a.emailVerified AS isVerified,
+        TRUE AS isPrimary,
+        a.createdAt AS verifiedAt,
+        a.createdAt
+    FROM
+        accounts a
+    WHERE
+        uid = LOWER(inUid))
+
+    UNION ALL
+
+    (SELECT
+        e.normalizedEmail,
+        e.email,
+        e.uid,
+        e.emailCode,
+        e.isVerified,
+        e.isPrimary,
+        e.verifiedAt,
+        e.createdAt
+    FROM
+        emails e
+    WHERE
+        uid = LOWER(inUid))
+    ORDER BY createdAt;
+END;
+
+CREATE PROCEDURE `deleteEmail_1` (
+    IN `inUid` BINARY(16),
+    IN `inNormalizedEmail` VARCHAR(255)
+)
+BEGIN
+    SET @deletedCount = 0;
+    SET @isPrimary = 0;
+
+    -- Don't delete any email that are considered primary emails.
+    -- Since this should be the most common scenario, we try it first
+    DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND uid = inUid AND isPrimary = 0;
+
+    -- Check to see if no rows were deleted, if this is the case, check to see if user
+    -- is attempting to deleted the email in the account table.
+    SELECT ROW_COUNT() INTO @deletedCount;
+    IF @deletedCount = 0 THEN
+      SELECT COUNT(*) INTO @isPrimary FROM accounts WHERE normalizedEmail = inNormalizedEmail AND uid = inUid;
+
+      IF @isPrimary = 1 THEN
+        SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
```

Comment:
Does this also need a `SET MYSQL_ERRNO = 1062`? Or I suppose a different number since it's not creating a duplicate...